README.md
# flutter_auth_code_textfield

A custom auth code textfield widget.

一个Flutter自定义的验证码Widget,纯Flutter代码，支持安卓，iOS.

## Useage

Mode: singleItem , like Keep 

Item模式，如Keep的风格。

```
AuthCodeTextfield(
                  mode: AuthCodeMode.singleItem,
                  itemWidth: 50,
                  itemHeight: 50,
                  itemSpacing: 35 * scaleWidth,
                  itemBackgroundColor: Color.fromRGBO(120, 114, 127, 1),
                  textColor: Colors.white,
                  cursorColor: Color.fromRGBO(94,178,138,1),
                  onChanged: (s) {
                    setState(() {
                      _inputText = s;
                    });
                  },
                ),
```


iOS            |  Android
:-------------------------:|:-------------------------:
![](https://github.com/JerryFans/flutter_auth_code_textfield/raw/master/mode1.png)  |  ![](https://github.com/JerryFans/flutter_auth_code_textfield/raw/master/mode3.png)

Mode: bottomLine , like DingDing 

bottomLine模式，如钉钉的风格。

```
AuthCodeTextfield(
                  mode: AuthCodeMode.bottomLine,
                  itemWidth: 50,
                  itemHeight: 60,
                  itemSpacing: 35 * scaleWidth,
                  itemBottomLineColor: Color.fromRGBO(229, 231, 233, 1),
                  cursorWidth: 1,
                  cursorHeight: 30,
                  onChanged: (s) {
                    _inputText = s;
                    if (s.toString().length >= 4 && widget.onSubmited != null) {
                      widget.onSubmited(_inputText);
                      Navigator.pop(context);
                    }
                  },
                ),
```

iOS            |  Android
:-------------------------:|:-------------------------:
![](https://github.com/JerryFans/flutter_auth_code_textfield/raw/master/mode2.png)  |  ![](https://github.com/JerryFans/flutter_auth_code_textfield/raw/master/mode5.png)



## Getting Started

Add this to your package's pubspec.yaml file:

dependencies:
  auth_code_textfield: ^0.0.2

If you want to support flutter 2.0 null safety please:

dependencies:
  auth_code_textfield: ^2.0.0

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.0.1] - (2020.5.6)

发布第一个版本

## [0.0.2] - (2020.5.6)

完善一些日志信息，删除一些不必要的注释

## [2.0.0] - (2021.9.21)

support 2.0 null safety

## [2.0.1] - (2022.2.21)

support border colod and border width

## [2.0.2] - (2022.7.14)

add cursor bottom offset property, if null, default cursor pos is center vertical

## [2.0.3] - (2022.7.14)

add fontweight property

## [2.0.4] - (2022.7.17)

add autofocus,controller,focusNode property

## [2.0.5] - (2022.7.18)

fix some issue when add controller property.

## [2.0.6] - (2022.7.18)

fix some issue


test/auth_code_textfield.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:auth_code_textfield/auth_code_textfield.dart';

void main() {
  
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'item_login_page.dart';
import 'bottom_line_login_page.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);
  final String? title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _receiveCode = '';
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        brightness: Brightness.dark,
        title: Text(widget.title ?? ""),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text('Receive code : $_receiveCode'),
            SizedBox(height: 20),
            Container(
              width: 200,
              height: 44,
              child: ElevatedButton(
                onPressed: () {
                  Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (_) => ItemLoginDemoPage(
                          onSubmited: (s) {
                            _receiveCode = s;
                            setState(() {});
                          },
                        ),
                      ));
                },
                child: Text('item mode demo'),
              ),
            ),
            SizedBox(height: 20),
            Container(
              width: 200,
              height: 44,
              child: ElevatedButton(
                onPressed: () {
                  Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (_) => BottomLineLoginPage(
                          onSubmited: (s) {
                            _receiveCode = s;
                            setState(() {});
                          },
                        ),
                      ));
                },
                child: Text('bottom line mode demo'),
              ),
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/auth_code_textfield.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'dart:async';

enum AuthCodeLength {
  four,
  six,
  eight,
}

enum AuthCodeMode {
  singleItem,
  bottomLine,
}

class AuthCodeTextfield extends StatefulWidget {
  final AuthCodeLength length;
  final AuthCodeMode mode;
  final double itemWidth;
  final double itemHeight;
  final double itemSpacing;
  final double itemBorderRadius;
  final double itemBottomLineWidth;
  final Color itemBackgroundColor;
  final Color itemBottomLineColor;
  final double cursorWidth; //光标宽，如果宽>高，会是一个水平的光标，如果宽<高，就是常规竖直的光标
  final double cursorHeight; //光标高
  final double? cursorBottomOffset; //cursor bottom offset, if nil default is center
  final Color cursorColor;
  final Color textColor;
  final FontWeight? fontWeight;
  final double fontSize;
  final double borderWidth;
  final Color borderColor;
  final ValueChanged? onChanged;
  final bool autofocus;
  final FocusNode? focusNode;
  final TextEditingController? controller;

  AuthCodeTextfield(
      {Key? key,
      this.length = AuthCodeLength.four,
      this.mode = AuthCodeMode.singleItem,
      this.itemWidth = 50.0,
      this.itemHeight = 50.0,
      this.itemSpacing = 1.0,
      this.itemBorderRadius = 4.0,
      this.itemBottomLineWidth = 1.0,
      this.itemBackgroundColor = Colors.transparent,
      this.itemBottomLineColor = Colors.blueGrey,
      this.cursorWidth = 20.0,
      this.cursorHeight = 2.0,
      this.cursorColor = Colors.blue,
      this.textColor = Colors.black,
      this.fontSize = 25,
      this.borderWidth = 0.0,
      this.borderColor = Colors.grey,
      this.onChanged, 
      this.cursorBottomOffset, 
      this.fontWeight, 
      this.autofocus = true, 
      this.focusNode, 
      this.controller})
      : super(key: key);

  @override
  _AuthCodeTextfieldState createState() => _AuthCodeTextfieldState();
}

class _AuthCodeTextfieldState extends State<AuthCodeTextfield> {
  String inputText = '';
  FocusNode _currentFocus = FocusNode();
  bool _visibleCursor = false;
  Timer? _timer;

  List<Widget> getTextListWidget() {
    List<Widget> list = [];
    for (int i = 0; i < inputText.length; i++) {
      var s = inputText[i];
      if (i != 0) {
        list.add(
          SizedBox(
            width: widget.itemSpacing,
          ),
        );
      }
      list.add(
        Container(
          alignment: Alignment.center,
          width: widget.itemWidth,
          height: widget.itemHeight,
          child: Text(
            s,
            style:
                TextStyle(fontSize: widget.fontSize, color: widget.textColor, fontWeight: widget.fontWeight),
          ),
        ),
      );
    }
    return list;
  }

  List<Widget> getItemListWidget() {
    List<Widget> list = [];
    for (int i = 0; i < getNumLenth(); i++) {
      if (i != 0) {
        list.add(
          SizedBox(
            width: widget.itemSpacing,
          ),
        );
      }
      list.add(
        Container(
          decoration: BoxDecoration(
            color: widget.itemBackgroundColor,
            borderRadius: widget.mode == AuthCodeMode.singleItem
                ? BorderRadius.circular(widget.itemBorderRadius)
                : null,
            border: widget.mode == AuthCodeMode.bottomLine
                ? Border(
                    bottom: BorderSide(
                        color: widget.itemBottomLineColor,
                        width: widget.itemBottomLineWidth),
                  )
                : (widget.borderWidth > 0
                    ? Border(
                        bottom: BorderSide(
                            color: widget.borderColor,
                            width: widget.borderWidth),
                        top: BorderSide(
                            color: widget.borderColor,
                            width: widget.borderWidth),
                        left: BorderSide(
                            color: widget.borderColor,
                            width: widget.borderWidth),
                        right: BorderSide(
                            color: widget.borderColor,
                            width: widget.borderWidth),
                      )
                    : null),
          ),
          alignment: Alignment.center,
          width: widget.itemWidth,
          height: widget.itemHeight,
        ),
      );
    }
    return list;
  }

  int getNumLenth() {
    int length = 4;
    switch (widget.length) {
      case AuthCodeLength.four:
        length = 4;
        break;
      case AuthCodeLength.six:
        length = 6;
        break;
      case AuthCodeLength.eight:
        length = 8;
        break;
    }
    return length;
  }

  @override
  void initState() {
    _timer = Timer.periodic(Duration(milliseconds: 650), (Timer timer) {
      _visibleCursor = !_visibleCursor;
      setState(() {});
    });
    widget.controller?.addListener(() { 
      if (widget.controller?.text != null) {
        widget.onChanged?.call(widget.controller!.text);
        setState(() {
            inputText = widget.controller!.text;
        });
      }
    });
    super.initState();
  }

  @override
  void dispose() {
    _timer?.cancel();
    if (widget.focusNode == null) {
      _currentFocus.dispose();
    }
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    double mainWidth = widget.itemWidth * getNumLenth() +
        widget.itemSpacing * (getNumLenth() - 1);
    double cursorPaddingLeft = (widget.itemWidth - widget.cursorWidth) / 2 +
        (inputText.length * (widget.itemWidth + widget.itemSpacing));
    return Container(
      alignment: Alignment.center,
      width: mainWidth,
      height: widget.itemHeight,
      child: Stack(
        alignment: AlignmentDirectional.centerStart,
        children: <Widget>[
          Row(
            mainAxisAlignment: MainAxisAlignment.start,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: getItemListWidget(),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.start,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: getTextListWidget(),
          ),
          Center(
            child: TextField(
              enableInteractiveSelection: false,
              controller: widget.controller,
              autofocus: widget.autofocus,
              focusNode: widget.focusNode != null ? widget.focusNode : _currentFocus,
              showCursor: false,
              inputFormatters: <TextInputFormatter>[
                FilteringTextInputFormatter.digitsOnly,
                LengthLimitingTextInputFormatter(getNumLenth())
              ],
              decoration: InputDecoration(
                border: InputBorder.none,
                counterText: '',
              ),
              maxLength: getNumLenth(),
              keyboardType: TextInputType.number,
              style: TextStyle(
                fontSize: 16.0,
                color: Colors.transparent,
              ),
              onChanged: (s) {
                if (widget.controller != null) {
                  return;
                }
                widget.onChanged?.call(s);
                setState(() {
                  inputText = s;
                });
              },
            ),
          ),
          Positioned(
            bottom: widget.cursorBottomOffset,
            child: Visibility(
              visible: _visibleCursor,
              child: Padding(
                padding: EdgeInsets.only(left: cursorPaddingLeft),
                child: Container(
                  width: widget.cursorWidth,
                  height: widget.cursorHeight,
                  decoration: BoxDecoration(
                    color: widget.cursorColor,
                    borderRadius: BorderRadius.circular(widget.cursorWidth / 2),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/bottom_line_login_page.dart
import 'package:flutter/material.dart';
import 'package:auth_code_textfield/auth_code_textfield.dart';
import 'package:flutter/services.dart';

class BottomLineLoginPage extends StatefulWidget {
  final ValueChanged? onSubmited;
  BottomLineLoginPage({Key? key, this.onSubmited}) : super(key: key);

  @override
  _BottomLineLoginPageState createState() => _BottomLineLoginPageState();
}

class _BottomLineLoginPageState extends State<BottomLineLoginPage> {
  String _inputText = '';

  @override
  void initState() {
    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.dark);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    double scaleWidth = MediaQuery.of(context).size.width / 414;
    return Scaffold(
      appBar: AppBar(
        brightness: Brightness.light,
        backgroundColor: Colors.white,
        elevation: 0,
        leading: IconButton(
          onPressed: () {
            Navigator.pop(context);
          },
          icon: Icon(Icons.arrow_back_ios),
          color: Colors.black,
        ),
      ),
      backgroundColor: Colors.white,
      body: Stack(
        alignment: Alignment.center,
        children: [
          Center(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  '验证码已发送至手机：',
                  style: TextStyle(
                    fontSize: 22,
                    color: Colors.black,
                  ),
                ),
                Text(
                  '+86 158XXXXXXXX',
                  style: TextStyle(
                    fontSize: 15,
                    color: Colors.black,
                  ),
                ),
                SizedBox(
                  height: 25,
                ),
                Text(
                  '请输入验证码',
                  style: TextStyle(
                    fontSize: 12,
                    color: Color.fromRGBO(28, 39, 51, 1),
                  ),
                ),
                AuthCodeTextfield(
                  mode: AuthCodeMode.bottomLine,
                  itemWidth: 50,
                  itemHeight: 60,
                  itemSpacing: 35 * scaleWidth,
                  itemBottomLineColor: Color.fromRGBO(229, 231, 233, 1),
                  cursorWidth: 1,
                  cursorHeight: 30,
                  onChanged: (s) {
                    _inputText = s;
                    if (s.toString().length >= 4 && widget.onSubmited != null) {
                      widget.onSubmited?.call(_inputText);
                      Navigator.pop(context);
                    }
                  },
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/item_login_page.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:auth_code_textfield/auth_code_textfield.dart';

class ItemLoginDemoPage extends StatefulWidget {
  final ValueChanged? onSubmited;
  ItemLoginDemoPage({Key? key, this.onSubmited}) : super(key: key);

  @override
  _ItemLoginDemoPageState createState() => _ItemLoginDemoPageState();
}

class _ItemLoginDemoPageState extends State<ItemLoginDemoPage> {
  String _inputText = '';
  TextEditingController controller = TextEditingController();

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    double scaleWidth = MediaQuery.of(context).size.width / 414;
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Color.fromRGBO(87, 79, 95, 1),
        elevation: 0,
      ),
      backgroundColor: Color.fromRGBO(87, 79, 95, 1),
      body: Stack(
        alignment: Alignment.center,
        children: [
          Center(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  '输入验证码',
                  style: TextStyle(
                      fontSize: 26,
                      color: Colors.white,
                      fontWeight: FontWeight.w600),
                ),
                SizedBox(
                  height: 25,
                ),
                RichText(
                  text: TextSpan(
                    text: '已发送4位验证码至',
                    style: TextStyle(fontSize: 15, color: Colors.white),
                    children: <TextSpan>[
                      TextSpan(
                        text: ' +86 158XXXXXXXX',
                        style: TextStyle(
                            fontSize: 15,
                            color: Color.fromRGBO(168, 162, 172, 1)),
                      ),
                    ],
                  ),
                ),
                SizedBox(
                  height: 25,
                ),
                AuthCodeTextfield(
                  mode: AuthCodeMode.singleItem,
                  controller: controller,
                  itemWidth: 50,
                  itemHeight: 50,
                  itemSpacing: 35 * scaleWidth,
                  itemBackgroundColor: Color.fromRGBO(120, 114, 127, 1),
                  textColor: Colors.white,
                  cursorColor: Color.fromRGBO(94, 178, 138, 1),
                  onChanged: (s) {
                    setState(() {
                      _inputText = s;
                    });
                    if (s.length == 4) {
                      controller.text = '';
                    }
                  },
                ),
                SizedBox(
                  height: 25,
                ),
                Container(
                  width: 50 * 4 + (35 * scaleWidth * 3),
                  height: 44,
                  child: CupertinoButton(
                    disabledColor: Color.fromRGBO(54, 127, 105, 1),
                    color: Color.fromRGBO(33, 191, 125, 1),
                    padding: EdgeInsets.zero,
                    borderRadius: BorderRadius.all(Radius.circular(22)),
                    onPressed: _inputText.length >= 4
                        ? () {
                            if (widget.onSubmited != null) {
                              widget.onSubmited?.call(_inputText);
                            }
                            Navigator.pop(context);
                          }
                        : null,
                    child: Text(
                      '确定',
                      style: TextStyle(fontSize: 17, color: Colors.white),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
